File: src/breslin/clientside/game/game.cpp
//header
#include "game.h"

//log
#include "../tdreamsock/dreamSockLog.h"

//applicationBreslin
#include "../application/applicationBreslin.h"

//network
#include "../network/network.h"

//shape
#include "../shape/shape.h"

//bytebuffer
#include "../bytebuffer/byteBuffer.h"

//ObjectTitle
#include "../billboard/objectTitle.h"

#include "../../fsm/stateMachine.h"

template <class entity_type> class State;

#include "states/gameStates.h"

//rotation
#include "../rotation/rotation.h"

//move
#include "../move/move.h"

/***************************************
*			          CONSTRUCTORS
***************************************/
Game::Game(ApplicationBreslin* application)
{
	mApplication = application;

	mShapeVector = new std::vector<Shape*>();
	mShapeGhostVector = new std::vector<Shape*>();

	//keys
	mKeyUp = 1;
	mKeyDown = 2;
	mKeyLeft = 4;
	mKeyRight = 8;
	mKeyCounterClockwise = 16;
	mKeyClockwise = 32;

	//input
	mKeyCurrent = 0;
	mKeyLast = 0;
	
	mControlObject = NULL;

	//time
    	mRunNetworkTime = 0.0f;
	mFrameTimeServer = 0.0f;

	//set Camera
	// Position it at 500 in Z direction
    	mApplication->getCamera()->setPosition(Ogre::Vector3(0,20,20));
    	// Look back along -Z
    	mApplication->getCamera()->lookAt(Ogre::Vector3(0,0,0));
    	mApplication->getCamera()->setNearClipDistance(5);

	createScene();

	// states
        mStateMachine =  new StateMachine<Game>(this);
        mStateMachine->setCurrentState      (PLAY_GAME::Instance());
        mStateMachine->setPreviousState     (PLAY_GAME::Instance());
        mStateMachine->setGlobalState       (NULL);
}

Game::~Game()
{
}

void Game::reset()
{
	//set mKeyArray to false
        for (int i = 0; i < 128; i++)
        {
                mApplication->mKeyArray[i] = false;
        }
        mKeyCurrent = 0;
}
void Game::processMoveControls()
{
       	//process input
        mKeyCurrent = 0;

        //UP
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_W))
        {
                mApplication->mKeyArray[119] = true;
                mKeyCurrent |= mKeyUp;
        }
        else
        {
                mApplication->mKeyArray[119] = false;
        }

        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_UP))
        {
                mApplication->mKeyArray[119] = true;
                mKeyCurrent |= mKeyUp;
        }
        else
        {
                mApplication->mKeyArray[119] = false;
        }

        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_NUMPAD8))
        {
                mApplication->mKeyArray[119] = true;
                mKeyCurrent |= mKeyUp;
        }
        else
        {
                mApplication->mKeyArray[119] = false;
        }
        
	//DOWN
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_S))
        {
                mApplication->mKeyArray[115] = true;
                mKeyCurrent |= mKeyDown;
        }
        else
        {
                mApplication->mKeyArray[115] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_DOWN))
        {
                mApplication->mKeyArray[115] = true;
                mKeyCurrent |= mKeyDown;
        }
        else
        {
                mApplication->mKeyArray[115] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_NUMPAD2))
        {
                mApplication->mKeyArray[115] = true;
                mKeyCurrent |= mKeyDown;
        }
        else
        {
                mApplication->mKeyArray[115] = false;
        }
 	
	//LEFT
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_A))
        {
                mApplication->mKeyArray[97] = true;
                mKeyCurrent |= mKeyLeft;
        }
        else
        {
                mApplication->mKeyArray[97] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_LEFT))
        {
                mApplication->mKeyArray[97] = true;
                mKeyCurrent |= mKeyLeft;
        }
        else
        {
                mApplication->mKeyArray[97] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_NUMPAD4))
        {
                mApplication->mKeyArray[97] = true;
                mKeyCurrent |= mKeyLeft;
        }
        else
        {
                mApplication->mKeyArray[97] = false;
        }

        //RIGHT
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_D))
        {
                mApplication->mKeyArray[100] = true;
                mKeyCurrent |= mKeyRight;
        }
        else
        {
                mApplication->mKeyArray[100] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_RIGHT))
        {
                mApplication->mKeyArray[100] = true;
                mKeyCurrent |= mKeyRight;
        }
        else
        {
                mApplication->mKeyArray[100] = false;
        }
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_NUMPAD6))
        {
                mApplication->mKeyArray[100] = true;
                mKeyCurrent |= mKeyRight;
        }
        else
        {
                mApplication->mKeyArray[100] = false;
        }
   	
	//COUNTERCLOCKWISE
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_Z))
        {
                mApplication->mKeyArray[122] = true;
                mKeyCurrent |= mKeyCounterClockwise;
        }
        else
        {
                mApplication->mKeyArray[122] = false;
        }

        //CLOCKWISE
        if (mApplication->getKeyboard()->isKeyDown(OIS::KC_X))
        {
                mApplication->mKeyArray[120] = true;
                mKeyCurrent |= mKeyClockwise;
        }
        else
        {
                mApplication->mKeyArray[120] = false;
        }

        //network outgoing
        sendByteBuffer();


}

//i am guessing i am not clearing the shape arrray???
void Game::remove()
{
/*
        if (mShapeVector)
        {
                for (unsigned int i = 0; i < mShapeVector->size(); i++)
                {
			Shape* shape = mShapeVector->at(i);

        		//delete objectTitles
			if (shape)
			{
        			delete shape->mGhost->mObjectTitle;
        			delete shape->mObjectTitle;
        			delete shape->mGhost;
        			delete shape;
			}
                }
        }
	
	mApplication->getSceneManager()->destroyAllEntities();
        mApplication->mSceneMgr->destroyLight(mPointLight);

	//clear the vectors....	
	mShapeVector->clear();
	mShapeGhostVector->clear();
*/
}

/*********************************
		Update
**********************************/
void Game::processUpdate()
{
	mStateMachine->update();

	for (unsigned int i = 0; i < mShapeVector->size(); i++)
	{
		mShapeVector->at(i)->interpolateTick(mApplication->getRenderTime());
	}
}

/*********************************
*		SHAPE
**********************************/
void Game::addShape(ByteBuffer* byteBuffer)
{
	byteBuffer->BeginReading();
        byteBuffer->ReadByte(); //should read -103 to add a shape..
        int local  =    byteBuffer->ReadByte();
        int index  =    byteBuffer->ReadShort();

	bool exists = false;	
	for (unsigned int i = 0; i < mShapeVector->size(); i++)
	{
		if (mShapeVector->at(i)->mIndex == index)
		{
			exists = true;
			if (local)
			{
				mShapeVector->at(i)->mLocal = true;
			}
			else
			{
				mShapeVector->at(i)->mLocal = false;
			}
		}
	}
	
	if (!exists)
	{
		Shape* shape = new Shape(mApplication,byteBuffer,false);  //you should just need to call this...
		shape->mRotation = new Rotation(shape);
		shape->mMove = new Move(shape);

		//put shape and ghost in game vectors so they can be looped and game now knows of them.
		mShapeVector->push_back(shape);
		mShapeGhostVector->push_back(shape->mGhost);
	}
}

void Game::removeShape(ByteBuffer* byteBuffer)
{
	//get the index which is the mIndex...
	int index = byteBuffer->ReadByte();
	
	//let's loop so we can also get index of vector where shape resides so we can erase it as well as deleting other parts of shape.
	for (unsigned int i = 0; i < mShapeVector->size(); i++)
	{
		if (mShapeVector->at(i)->mIndex == index)
		{
			Shape* shape = mShapeVector->at(i);
			
			//delete objectTitles
 			delete shape->mGhost->mObjectTitle;
 			delete shape->mObjectTitle;

			//delete entities by name
 			mApplication->getSceneManager()->destroyEntity(shape->mGhost->mName);
 			mApplication->getSceneManager()->destroyEntity(shape->mName);

			//erase shapes from vectors
			mShapeVector->erase(mShapeVector->begin()+i);
			mShapeGhostVector->erase(mShapeGhostVector->begin()+i);
		}
	}
}

/*
        if (mShapeVector)
        {
                for (unsigned int i = 0; i < mShapeVector->size(); i++)
                {
                        delete mShapeVector->at(i);
                        delete mShapeGhostVector->at(i);
                }
        }
*/

Shape* Game::getShape(int id)
{
	Shape* shape = NULL;

	for (unsigned int i = 0; i < mShapeVector->size(); i++)
	{
		Shape* curShape = mShapeVector->at(i);
		if (curShape->mIndex == id)
		{
			shape = curShape;
		}
	}

	if(!shape)
	{
		return NULL;
	}
	else
	{
		return shape;
	}
}

void Game::createScene()
{
        mApplication->mSceneMgr->setAmbientLight(Ogre::ColourValue(0.75, 0.75, 0.75));

        mPointLight = mApplication->mSceneMgr->createLight("pointLight");
        mPointLight->setType(Ogre::Light::LT_POINT);
        mPointLight->setPosition(Ogre::Vector3(250, 150, 250));
        mPointLight->setDiffuseColour(Ogre::ColourValue::White);
        mPointLight->setSpecularColour(Ogre::ColourValue::White);

	// create a floor mesh resource
        MeshManager::getSingleton().createPlane("floor", ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME,
               Plane(Vector3::UNIT_Y, 0), 500, 500, 10, 10, true, 1, 10, 10, Vector3::UNIT_Z);

        mFloor = mApplication->mSceneMgr->createEntity("Floor", "floor");
        mFloor->setMaterialName("Examples/Rockwall");
        mFloor->setCastShadows(false);
	mFloorNode = mApplication->mSceneMgr->getRootSceneNode()->createChildSceneNode();
	mFloorNode->attachObject(mFloor);
	mFloorNode->setPosition(0,-13.0f,0);


	// create a northwall mesh resource
        MeshManager::getSingleton().createPlane("northwall", ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME,
               Plane(Vector3::UNIT_Z, 0), 500, 50, 10, 10, true, 1, 10, 10, Vector3::UNIT_Y);

        mNorthWall = mApplication->mSceneMgr->createEntity("Northwall", "northwall");
        mNorthWall->setMaterialName("Examples/Rockwall");
        mNorthWall->setCastShadows(false);
	mNorthWallNode = mApplication->mSceneMgr->getRootSceneNode()->createChildSceneNode();
	mNorthWallNode->attachObject(mNorthWall);
	mNorthWallNode->setPosition(250,0,0);

	// create a eastwall mesh resource
        MeshManager::getSingleton().createPlane("eastwall", ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME,
               Plane(Vector3::UNIT_X, 0), 500, 50, 10, 10, true, 1, 10, 10, Vector3::UNIT_Y);

        mEastWall = mApplication->mSceneMgr->createEntity("Eastwall", "eastwall");
        mEastWall->setMaterialName("Examples/Rockwall");
        mEastWall->setCastShadows(false);
	mEastWallNode = mApplication->mSceneMgr->getRootSceneNode()->createChildSceneNode();
	mEastWallNode->attachObject(mEastWall);
	mEastWallNode->setPosition(500,0,250);



}

/*********************************
		Network
**********************************/

void Game::checkByteBuffer(ByteBuffer* byteBuffer)
{
	byteBuffer->BeginReading();

        int type = byteBuffer->ReadByte();

        switch(type)
        {
        	case mMessageAddShape:
                	addShape(byteBuffer);
                        break;

                case mMessageRemoveShape:
                 	removeShape(byteBuffer);
                        break;

                case mMessageFrame:
                       	if (!mApplication->mNetwork->mIgnorePacket)
			{
                        	readServerTick(byteBuffer);
                        }
                        break;
		case mMessageReportStandings:
 			setStandings(byteBuffer);
                        break;
		case mMessageSetText:
        		int index = byteBuffer->ReadShort();
   			for (unsigned int i = 0; i < mShapeVector->size(); i++)
                	{
                        	Shape* shape = mShapeVector->at(i);
				if (shape->mIndex == index)
				{
					shape->setText(byteBuffer);	
				}
			}
			break;
	}
}

void Game::readServerTick(ByteBuffer* byteBuffer)
{
        // Skip sequences
        byteBuffer->ReadShort();
        mFrameTimeServer = byteBuffer->ReadByte();

        while (byteBuffer->getReadCount() <= byteBuffer->GetSize())
        {
                mApplication->mDetailsPanel->setParamValue(11, Ogre::StringConverter::toString(byteBuffer->GetSize()));

                int id = byteBuffer->ReadByte();

                Shape* shape = NULL;
                shape = getShape(id);

		if (shape)
                {
                        shape->processDeltaByteBuffer(byteBuffer);
                }
                else
                {
                        //LogString("INVALID SHAPE ID");
                }
        }
}

void Game::sendByteBuffer()
{
	mRunNetworkTime += mApplication->getRenderTime() * 1000.0f;

    	// Framerate is too high
    	if(mRunNetworkTime > (1000 / 60))
    	{
        	// Build delta-compressed move command
        	int flags = 0;

        	//if key has not been changed return having done nothing 
        	if(mKeyLast != mKeyCurrent)
        	{
                	flags |= mMessageKey;
        	}
		else
		{
			return;
		}

		//create byteBuffer
		ByteBuffer* byteBuffer = new ByteBuffer();

		//WRITE: type
		byteBuffer->WriteByte(mMessageFrame);

                //WRITE: key
               	byteBuffer->WriteByte(mKeyCurrent);

        	//set 'last' commands for diff
        	mKeyLast = mKeyCurrent;

        	// Send the packet
        	mApplication->mNetwork->send(byteBuffer);

		mRunNetworkTime = 0.0f;
    	}
}

void Game::setStandings(ByteBuffer* byteBuffer)
{
        mApplication->mStringReset.clear();

        int length = byteBuffer->ReadByte();
        for (int i = 0; i < length; i++)
        {
                char c =  byteBuffer->ReadByte();
                mApplication->mStringReset.append(1,c);
        }
        if (mApplication->mLabelReset && mApplication->mStringReset.size() > 0)
        {
                mApplication->mLabelReset->setCaption(mApplication->mStringReset);
		LogString("got someting for standings");
        }
        else
        {
                LogString("no label or no string for standings");
        }
}

File: src/breslin/clientside/game/game.php
var Game = new Class(
{

initialize: function(application)
{
	this.mDelayClick = 100;
	this.mDelayClickCounter = 0;
	
	//keys
 	this.mKeyArray     = new Array();

	// constants
	this.mCommandKey          = 1;
	this.mCommandFrameTime = 2;

	this.mSequence = 0;

	this.mMessageFrame = 2;  //changed this to browser code

	this.mMessageAddShape    = -103;
	this.mMessageRemoveShape = -104;

	//application	
	this.mApplication = application;	

	//shapes
	this.mShapeVector      = new Array();
	this.mShapeGhostVector = new Array();
	this.mControlObject      = 0; 
	this.mControlObjectGhost = 0; 
	this.mOffSet = new Vector3D();

	//byteBuffer
	this.mByteBuffer = new ByteBuffer();

        //keys
        this.mKeyUp = 1;
        this.mKeyDown = 2;
        this.mKeyLeft = 4;
        this.mKeyRight = 8;
        this.mKeyCounterClockwise = 16;
        this.mKeyClockwise = 32;

        //input
        this.mKeyCurrent = 0;
        this.mKeyLast = 0;

        //time
        this.mRunNetworkTime = 0.0;
	this.mPollDelay = 100;
	this.mPollDelayCounter = 0;
	this.mFrameTimeServer = 0;

	this.mStateMachine = new StateMachine(this);
        this.mPLAY_GAME = new PLAY_GAME(this);
        this.mRESET_GAME = new RESET_GAME(this);
        this.mStateMachine.setGlobalState(0);
        this.mStateMachine.changeState(this.mPLAY_GAME);
/*
        //set Camera
        // Position it at 500 in Z direction
        mApplication->getCamera()->setPosition(Ogre::Vector3(0,20,20));
        // Look back along -Z
        mApplication->getCamera()->lookAt(Ogre::Vector3(0,0,0));
        mApplication->getCamera()->setNearClipDistance(5);
*/
},


log: function(msg)
{
	setTimeout(function()
        {
        	throw new Error(msg);
        }, 0);
},

remove: function()
{
	//set shapes to not visible and clear shapeArrays

	if (this.mShapeGhostVector)
	{
		for (i = 0; i < this.mShapeGhostVector.length; i++)
		{
			this.mShapeGhostVector[i].mDiv.mDiv.removeChild(this.mShapeGhostVector[i].mObjectTitle);
			this.mShapeGhostVector[i].mDiv.mDiv.removeChild(this.mShapeGhostVector[i].mMesh);
			document.body.removeChild(this.mShapeGhostVector[i].mDiv.mDiv);
		}
		//set vector to 0 length...
		this.mShapeGhostVector.length = 0;
	}
	if (this.mShapeVector)
	{
		for (i = 0; i < this.mShapeVector.length; i++)
		{
			this.mShapeVector[i].mDiv.mDiv.removeChild(this.mShapeVector[i].mObjectTitle);
			this.mShapeVector[i].mDiv.mDiv.removeChild(this.mShapeVector[i].mMesh);
			document.body.removeChild(this.mShapeVector[i].mDiv.mDiv);
		}
		//set vector to 0 length...
		this.mShapeVector.length = 0;
	}

},

/*********************************
                Update
**********************************/
processUpdate: function()
{
	//set intervalCount back to zero to signify we are still running game loop
	mIntervalCount = 0;

	//set the offset
	if (this.mControlObject)
	{
		x = parseFloat(this.mControlObject.mPosition.x);
		z = parseFloat(this.mControlObject.mPosition.z);
		//x = parseFloat(x * 2);
		//z = parseFloat(z * 2);

		this.mOffSet.x = parseFloat(this.mApplication.mScreenCenter.x) - parseFloat(x);
        	this.mOffSet.y = 0;
        	this.mOffSet.z = parseFloat(this.mApplication.mScreenCenter.z) - parseFloat(z);
	}

	this.mStateMachine.update();

	for (i = 0; i < this.mShapeVector.length; i++)
	{
		this.mShapeVector[i].interpolateTick(this.mApplication.getRenderTime());
	}
},

/*********************************
               NETWORK 
**********************************/
checkForByteBuffer: function()
{

},

/*********************************
*               SHAPE
**********************************/
addShape: function(byteBuffer)
{
	shape = new Shape(this.mApplication,byteBuffer,false);

	shape.mMove = new Move(shape);

	//put shape and ghost in game vectors so they can be looped and game now knows of them.
        this.mShapeVector.push(shape);
        this.mShapeGhostVector.push(shape.mGhost);
/*
	if (shape.mLocal == 1)
	{
		shape.mGhost.setVisible(true);
	}
	else
	{
		shape.mGhost.setVisible(false);
	}
*/
	shape.mGhost.setVisible(false);

},

removeShape: function(byteBuffer)
{
	if (this.mApplication.mLeaveGame || this.mApplication.mSentLeaveGame)
	{
	}
	else
	{
 		byteBuffer.beginReading();
        	type     = byteBuffer.readByte();
        	clientID = byteBuffer.readByte();
        	index    = byteBuffer.readByte();
	
  		for (i=0; i < this.mShapeVector.length; i++)
        	{
                	if (index == this.mShapeVector[i].mIndex)
			{
 				this.mShapeVector[i].mDiv.mDiv.removeChild(this.mShapeVector[i].mObjectTitle);
                        	this.mShapeVector[i].mDiv.mDiv.removeChild(this.mShapeVector[i].mMesh);
                        	document.body.removeChild(this.mShapeVector[i].mDiv.mDiv);

				this.mShapeVector.splice(i,1);
			}
			inverseIndex = index * -1;
                	if (inverseIndex == this.mShapeGhostVector[i].mIndex)
			{
 				this.mShapeGhostVector[i].mDiv.mDiv.removeChild(this.mShapeGhostVector[i].mObjectTitle);
                        	this.mShapeGhostVector[i].mDiv.mDiv.removeChild(this.mShapeGhostVector[i].mMesh);
                        	document.body.removeChild(this.mShapeGhostVector[i].mDiv.mDiv);

				this.mShapeGhostVector.splice(i,1);
			}
		}
	}
},

readServerTick: function(byteBuffer)
{
	this.mApplication.mIntervalCountLast = this.mApplication.mIntervalCount;
	this.mApplication.mIntervalCount++;

	seq = byteBuffer.readByte(); //seq
	this.mSequence = seq;
	this.mFrameTimeServer = byteBuffer.readByte(); //time
	
	while (byteBuffer.mReadCount < byteBuffer.mBufferArray.length)
	{
		var id = byteBuffer.readByte();
		
		var shape = this.getShape(id);
		
		if (shape != 0)
		{
			shape.processDeltaByteBuffer(byteBuffer);
		}
		else
		{
			//do nothing
		}
	}
},

setText: function(byteBuffer)
{
	var text = '';
        var index = byteBuffer.readByte();
        var length = byteBuffer.readByte();
        
	text = text + byteBuffer.readByte();
/*
	for (i = 0; i < length; i++)
        {
                text = text + byteBuffer.readByte();
        }
*/
	var shape = this.getShape(index);
	shape.setText(text);
},

/*
askQuestion: function(byteBuffer)
{
        this.mApplicationPartido.mStringQuestion = '';
        var length = byteBuffer.readByte();

        for (i = 0; i < length; i++)
        {
                this.mApplicationPartido.mStringQuestion = this.mApplicationPartido.mStringQuestion + byteBuffer.readByte();
        }

        if (this.mApplicationPartido.mLabelQuestion)
        {
                this.mApplicationPartido.mLabelQuestion.value = this.mApplicationPartido.mStringQuestion;
        }
        else
        {
        }

        //reset mAnswerTime
        this.mApplicationPartido.mAnswerTime = 0;
}


*/
getShape: function(id)
{
	var shape = 0;

	for (i=0; i < this.mShapeVector.length; i++)
	{
		var curShape = this.mShapeVector[i];	
		if (curShape.mIndex == id)
		{
			shape = curShape;
		}
	}
	if (shape == 0)
	{
		return 0;
	}
	else
	{
		return shape;
	}
},

sendByteBuffer: function()
{
        this.mRunNetworkTime += this.mApplication.getRenderTime() * 1000.0;
  	
	// Framerate is too high
        if(this.mRunNetworkTime > (1000 / 60))
        {
                // Build delta-compressed move command
                flags = 0;

                //if key has not been changed return having done nothing
                if(this.mKeyLast != this.mKeyCurrent)
                {
                        flags |= this.mCommandKey;
                }
                else
                {
                        return;
                }

                // Send the packet
        	message = '1 ' + this.mKeyCurrent;
        	this.mApplication.mNetwork.mSocket.emit('send_move', message);

                //set 'last' commands for diff
                this.mKeyLast = this.mKeyCurrent;
                
		//reset network time so we can start count anew
		this.mRunNetworkTime = 0.0;
	}
},

/***************************************
*                       INPUT
******************************************/

processInput: function()
{
	this.mKeyCurrent = 0;

	if (this.mApplication.mKey_up)
	{
		this.mKeyCurrent |= this.mKeyUp;
	}
	if (this.mApplication.mKey_down)
	{
		this.mKeyCurrent |= this.mKeyDown;
	}
	if (this.mApplication.mKey_left)
	{
		this.mKeyCurrent |= this.mKeyLeft;
	}
	if (this.mApplication.mKey_right)
	{
		this.mKeyCurrent |= this.mKeyRight;
	}

	if (this.mApplication.mKey_counterclockwise)
	{
		this.mKeyCurrent |= this.mKeyCounterClockwise;
	}
	if (this.mApplication.mKey_clockwise)
	{
		this.mKeyCurrent |= this.mKeyClockwise;
	}
},

reset: function()
{

},

setStandings: function(byteBuffer)
{
	this.log('setStandings called');
        this.mApplication.mStringReset = '';

        this.mApplication.mStringReset = byteBuffer.readByte();
        this.log('mStringReset:' + this.mApplication.mStringReset);

        if (this.mApplication.mLabelReset)
        {
		this.log('setting mLabelReset!!');
                this.mApplication.mLabelReset.value = this.mApplication.mStringReset;
        }
        else
        {
        }
}

});
